syft/frameworks/torch/linalg/operations.py

```diff
@@ -63,6 +65,192 @@ def _ldl(t):
+        prec_frac = t.child.precision_fractional
+    elif isinstance(t, torch.Tensor):
+        t_type = "local"
+    else:
+        raise TypeError(
+            "The provided matrix should be a local torch.Tensor, a PointerTensor,",
+            "or an AdditiveSharedTensor",
+        )
+
+    if mode not in ["reduced", "complete", "r"]:
+        raise ValueError(
+            "mode should have one of the values in the list:" + str(["reduced", "complete", "r"])
+        )
+
+    ######## QR decomposition via Householder Reflection #########
+    n_rows, n_cols = t.shape
+
+    # Initiate R matrix from t
+    R = t.copy()
+
+    # Initiate identity matrix with size (n_rows, n_rows)
+    I = torch.diag(torch.Tensor([1.0] * n_rows))
+
+    # Send it to remote worker if t is pointer, secret share it if it's an AST
+    if t_type == "pointer":
+        I = I.send(location)
+    if t_type == "ast":
+        I = I.fix_prec(precision_fractional=prec_frac).share(*workers, crypto_provider=crypto_prov)
+
+    if not mode == "r":
+        # Initiate Q_transpose
+        Q_t = I.copy()
+
+    # Iteration via Household Reflection
+    for i in range(min(n_rows, n_cols)):
+
+        # Identity for this iteration, it has size (n_cols-i, n_cols-i)
+        I_i = I[i:, i:]
+
+        # Init 1st vector of the canonical base in the same worker as t
+        e = torch.zeros_like(t)[i:, 0].view(-1, 1)
+        e[0, 0] += 1
+
+        # Current vector in R to perform reflection
+        x = R[i:, i].view(-1, 1)
+
+        # Compute norm in MPC if it's an AST
+        x_norm = _norm_mpc(x, norm_factor) if t_type == "ast" else torch.sqrt(x.t() @ x).squeeze()
+
+        # Compute Householder transform
+        numerator = x @ x.t() - x_norm * (e @ x.t() + x @ e.t()) + (x.t() @ x) * (e @ e.t())
+        denominator = x.t() @ x - x_norm * x[0, 0]
+        # Need the line below to perform inverse of a number in MPC
+        inv_denominator = (0 * denominator + 1) / denominator
+        H = I_i - numerator * inv_denominator
+
+        # If it is not the 1st iteration
+        # expand matrix H with Identity at diagonal and zero elsewhere
+        if i > 0:
+            down_zeros = torch.zeros([n_rows - i, i])
+            up_zeros = torch.zeros([i, n_rows - i])
+
+            # Send them to remote worker if t is pointer, secret share it if its an AST
+            if t_type == "pointer":
+                down_zeros = down_zeros.send(t.child.location)
+                up_zeros = up_zeros.send(t.child.location)
+            if t_type == "ast":
+                down_zeros = down_zeros.fix_prec(precision_fractional=prec_frac).share(
+                    *workers, crypto_provider=crypto_prov
+                )
+                up_zeros = up_zeros.fix_prec(precision_fractional=prec_frac).share(
+                    *workers, crypto_provider=crypto_prov
+                )
+            left_cat = torch.cat((I[:i, :i], down_zeros), dim=0)
```

Comment:
Can you reuse `I_i` ?

Reply:
`I_i` is equal to `I[i:, i:]`, not to `I[:i, :i]`